---
title: 'MH600: Comparison of panel and WES results'
output:
  distill::distill_article:
    theme: hpstr
    toc: yes
    code_folding: true
    keep_md: yes
editor_options: 
  chunk_output_type: console
---

Panel -> (CNVkit+)PureCN results
WES -> CNVkit results

```{r setup, include=FALSE}
.packages <- c("tidyverse", "BiocManager", "gridExtra", "rmarkdown", "knitr", "reshape2", "ggpubr", "this.path", "yaml")
.bioconductor_packages <- c("GenomicRanges", "karyoploteR")

# Install CRAN packages (if not already installed)
.inst <- .packages %in% installed.packages()
if (length(.packages[!.inst]) > 0) install.packages(.packages[!.inst], dependencies = TRUE, quiet = TRUE, repos = "https://cran.uni-muenster.de/")

# Install Bioconductor packages (if not already installed)
.inst <- .bioconductor_packages %in% installed.packages()
if (length(.bioconductor_packages[!.inst]) > 0) BiocManager::install(.bioconductor_packages[!.inst], dependencies = TRUE, ask = FALSE)

# Load packages into session
lapply(.bioconductor_packages, require, character.only = TRUE)
lapply(.packages, require, character.only = TRUE)

dir <- dirname(this.path()) # Get directory of running file
```

# Panel design

```{r process bed file}
# read in yaml file
config <- read_yaml(paste0(dir, "/../config/config.yaml"))
bedfile <- paste0(dir, "/../", config$panel_design)
bed <- read.table(bedfile, header = FALSE, sep = "\t", skip = 2)
bed <- bed[, 1:4]
colnames(bed) <- c("chr", "start", "end", "gene")

# add "chr" to chromosome names if not present
bed$chr <- ifelse(grepl("^chr", bed$chr), bed$chr, paste0("chr", bed$chr))

markers <- bed
# from every gene, remove the pattern chr[0-9YX] and everything behind
markers$gene <- gsub(",*chr[0-9x]*:.*", "", markers$gene)
# remove rows with empty gene entries
markers <- markers[markers$gene != "", ]
# filter out only one location per gene
markers <- markers[!duplicated(markers[, c("chr", "gene")]), ]

# create granges object
gr <- makeGRangesFromDataFrame(bed, keep.extra.columns = TRUE)
```

## Coverage and gene density

Gene Density in blue, regions covered by panel design in red

```{r plot_karyotype, layout="l-page",fig.width=10, fig.height=6}
kp <- plotKaryotype(genome = "hg19", plot.type = 1, ideogram.plotter = NULL)
kpDataBackground(kp, data.panel = 2, col = "#eeeeee", r0 = -0.7, r1 = 0)
kpPlotRegions(kp, data = gr, col = "salmon", r0 = -0.695, r1 = 0)
kpPlotDensity(kp, gr, window.size = 0.6e6, data.panel = "ideogram", col = "#3388FF", border = "#3388FF", r0 = 1.2, r1 = 5)
```

For chromosome 17 only

```{r plot_only_chr17}
kp <- plotKaryotype(genome = "hg19", plot.type = 1, chromosome = "chr17", main = "Zoom in on chromosome 17")
# kpAddBaseNumbers(kp)
kpPlotMarkers(
  kp,
  chr = markers$chr,
  x = markers$start,
  labels = markers$gene,
  text.orientation = "vertical",
  marker.parts = c(0, 0.9, 0.1),
  line.color = "lightblue",
  label.color = "black",
  label.dist = 0.00001,
  max.iter = 100,
  cex = 0.9,
  adjust.label.position = TRUE,
  ignore.chromosome.ends = TRUE,
  r1 = 0.85,
  r0 = 0.2
)
kpDataBackground(kp, data.panel = 2, col = "#eeeeee", r0 = -0.7, r1 = -0.4)
kpPlotRegions(kp, data = gr, col = "salmon", r0 = -0.695, r1 = -0.4)
kpPlotDensity(kp, gr, window.size = 0.7e6, data.panel = "ideogram", col = "#3388FF", border = "#3388FF", r0 = 1.2, r1 = 2.5)
```

## Number of genes and gene density and total coverage per chromosome

```{r plot_genes_per_chr,fig.width=7, fig.height=8}
chr_lengths <- data.frame(
  chr = paste0("chr", c(1:22, "X", "Y")),
  chr_length = c(249, 242, 198, 191, 180, 171, 159, 145, 138, 134, 135, 133, 114, 107, 102, 90, 83, 80, 59, 64, 46, 50, 156, 57)
)

# filter those chromosomes that are not in the bed file
filtered_chr_lengths <- chr_lengths[chr_lengths$chr %in% unique(markers$chr), ]

bed %>%
  mutate(region_length = end - start) %>%
  group_by(chr) %>%
  summarise(total_length = sum(region_length)) %>%
  mutate(chromosome = gsub("chr", "", chr)) %>%
  mutate(chr_length = filtered_chr_lengths$chr_length) %>%
  mutate(perc_chr_coverage = total_length / 1e6 / chr_length * 1e2) -> bed_mod

markers %>%
  select(chr) %>%
  mutate(chromosome = gsub("chr", "", chr)) %>%
  group_by(chromosome) %>%
  summarise(number_of_genes = n()) %>%
  mutate(
    chr_length = filtered_chr_lengths$chr_length,
    gene_per_Kb_chr_length = number_of_genes / chr_length * 100
  ) %>%
  full_join(bed_mod, by = c("chromosome" = "chromosome")) %>%
  select(chromosome, number_of_genes, gene_per_Kb_chr_length, perc_chr_coverage) %>%
  arrange(match(chromosome, c("1", "2", "3", "4", "5", "6", "7", "8", "9", "10", "11", "12", "13", "14", "15", "16", "17", "18", "19", "20", "21", "22", "X", "Y")), desc(chromosome)) -> stats

ggarrange(
  stats %>% ggplot(aes(x = fct_inorder(chromosome), y = number_of_genes)) +
    geom_bar(stat = "identity", fill = "lightblue") +
    labs(title = "Number of genes per chromosome", x = "Chromosome", y = "Number of genes") +
    theme_minimal() +
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1)),
  stats %>% ggplot(aes(x = fct_inorder(chromosome), y = gene_per_Kb_chr_length)) +
    geom_bar(stat = "identity", fill = "salmon") +
    labs(title = "Gene density per chromosome", x = "Chromosome", y = "Genes per Kb total chromosome length") +
    theme_minimal() +
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1)),
  stats %>% ggplot(aes(x = fct_inorder(chromosome), y = perc_chr_coverage)) +
    geom_bar(stat = "identity", fill = "lightgreen") +
    labs(title = "Total coverage per chromosome by panel design", x = "Chromosome", y = "Total coverage (%)") +
    theme_minimal() +
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1)),
  ncol = 1,
  nrow = 3
)
```


# Number of successful sample runs


Out of the 36 panel samples (incl. discarded samples due to low mapping quality and read depth), the number of successful runs per PON and branch is as follows:

```{r}
read_table(paste0(dir, "/../results/stats/successful_purecn_runs.txt"),
  show_col_types = FALSE, col_names = c("number_of_samples", "branch")
) %>%
  filter(grepl("^results/purecn", branch)) %>%
  mutate(
    branch = gsub("/purecn/", "/purecn_no_pon/", branch),
    branch = gsub("results/purecn_", "", branch),
    branch = gsub("/$", "", branch),
    method = gsub(".*/", "", branch),
    pon = gsub("/.*", "", branch)
  ) %>%
  ggplot(aes(x = method, y = number_of_samples, fill = method)) +
  geom_col() +
  coord_flip() +
  geom_text(aes(label = number_of_samples), hjust = 1.1, colour = "white") +
  facet_wrap(~pon, )
```

Among these panel samples, 28 have matched WES samples. These are compared in the following.
~~Out of these 28 samples, the number of successful runs per branch is as follows:~~   

```{r, include = FALSE, eval = TRUE}
read_table(paste0(dir, "/../results/stats/successful_purecn_runs.txt"),
  show_col_types = FALSE, col_names = c("number_of_samples", "branch")
) %>%
  filter(grepl("collection", branch)) %>%
  mutate(
    branch = gsub("/purecn/", "/purecn_no_pon/", branch),
    branch = gsub("results/collection_of_specific_samples/", "", branch),
    branch = gsub("results/purecn_", "", branch),
    branch = gsub("/$", "", branch),
    method = gsub(".*/", "", branch),
    pon = gsub("/.*", "", branch)
  ) %>%
  ggplot(aes(x = method, y = number_of_samples, fill = method)) +
  geom_col() +
  coord_flip() +
  geom_text(aes(label = number_of_samples), hjust = 2, colour = "white") +
  facet_wrap(~pon)
```


# Compare and summarise all samples

```{r for MH collect all gene calls, include = FALSE, eval = TRUE}
# extract panel only genes
paneldesign_genes <- bed %>%
  select(gene) %>%
  unique()
cat("The paneldesign includes", paneldesign_genes %>% nrow(), "genes.\n")

panel_files <- list.files(paste0(dir, "/../results/collection_of_specific_samples"), pattern = "genes.csv", full.names = TRUE, recursive = TRUE)

read_csv(panel_files[1]) %>%
  select(gene.symbol) %>%
  filter(!grepl("Antitarget", gene.symbol)) -> purecn_genes
cat("The purecn results include", purecn_genes %>% nrow(), "genes.\n")

files <- list.files(paste0(dir, "/../resources/gold_standard"), pattern = "gene_call.txt", full.names = TRUE)
wes_list <- list()

for (file in files) {
  file %>%
    basename() %>%
    str_extract("\\.[A-Z0-9]*-") %>%
    str_remove("\\.") %>%
    str_remove("-") -> name

  read_tsv(file, show_col_types = FALSE) %>%
    arrange(Hugo_Symbol) %>%
    select(-Entrez_Gene_Id) %>%
    filter(Hugo_Symbol %in% purecn_genes$gene.symbol) %>%
    # merge rows with identical Hugo_Symbol entry
    group_by(Hugo_Symbol) %>%
    summarise_all(median, na.rm = TRUE) %>%
    # rename second column to the string saved in the variable name
    rename(!!name := 2) -> wes
  wes_list[[name]] <- wes
}

reduce(wes_list, full_join, by = "Hugo_Symbol") %>%
  # remove rows with only NAs in all but the first column
  filter(rowSums(is.na(.)) < ncol(.) - 1) %>%
  rename(gene = Hugo_Symbol) %>%
  {
    write_tsv(., paste0(dir, "/../results/stats/all_gene_calls_in_WES.tsv"))
  } -> wes_all

cat(wes_all %>% nrow(), "common genes in both panel and WES.\n")

write.table(wes_all$gene, paste0(dir, "/../results/stats/common_genes_in_panel_and_WES.txt"), row.names = FALSE, col.names = FALSE, quote = FALSE)
```

```{r}
files <- list.files(paste0(dir, "/../resources/gold_standard"), full.names = TRUE)
if (any(grepl("copywrit", files, ignore.case = TRUE))) {
  cat("The WES results are from Copywriter.\n")
} else if (any(grepl("cnvkit", files[1], ignore.case = TRUE))) {
  cat("The WES results are from CNVkit.\n")
}
```

```{r for panels collect all gene calls per branch, include = FALSE, eval = TRUE}
list.dirs(paste0(dir, "/../results/collection_of_specific_samples"), full.names = TRUE, recursive = FALSE) -> pons

# for every folder, read in all _genes.csv files, extract the gene and type columns, and merge into one table with gene as row names and C in the columns names according to the sample name contained in the file name. Then, for every branch, write the table to a file called branch+'_gene_calls.tsv'.

read_tsv(paste0(dir, "/../results/stats/all_gene_calls_in_WES.tsv")) %>%
  select(gene) %>%
  pull() -> all_genes

empty_row_tables <- c()
list <- list()
for (pon in pons) {
  list.dirs(path = pon, full.names = TRUE, recursive = FALSE) -> branches
  for (branch in branches) {
    list.files(path = branch, pattern = "_genes.csv", full.names = TRUE) -> files
    if (length(files) == 0) {
      next
    }
    list_branch <- list()
    for (file in files) {
      str_replace(file, "_genes.csv", "") %>%
        str_replace(".*/", "") -> name

      read_csv(file, show_col_types = FALSE) %>%
        select(gene.symbol, type) %>%
        filter(gene.symbol %in% all_genes) %>%
        rename(!!name := 2) -> foo

      if (foo %>% nrow() == 0) {
        empty_row_tables <- c(empty_row_tables, name)
      } else {
        list_branch[[name]] <- foo
      }
    }

    branch %>%
      str_remove(".*collection_of_specific_samples/") %>%
      str_replace("/", "_") -> branchname

    reduce(list_branch, full_join, by = "gene.symbol") %>%
      rename(gene = gene.symbol) %>%
      write_tsv(paste0(dir, "/../results/collection_of_specific_samples/", branchname, "_gene_calls.tsv"))
  }
}
```


```{r Create contingency tables for all samples}
# function to make sample name(s) similar (mainly for MH dataset)
rename_names <- function(names) {
  names <- gsub("M0*", "M", names)
  names <- gsub("-", "_", names)
  return(names)
}

wes_gene_calls <- paste0(dir, "/../results/stats/all_gene_calls_in_WES.tsv")
wes_different_names <- TRUE
# neutral = 2 # copy number value denoting no change in WES table
neutral <- 0 # copy number value denoting no change in WES table
matching_IDs <- paste0(dir, "/../resources/data/matching_IDs.tsv") # leave empty if identical names in panels

read_tsv(wes_gene_calls, col_names = TRUE, show_col_types = FALSE) -> wes
wes %>% arrange(gene) -> wes

# replace all NA values with 0
wes[is.na(wes)] <- 0

if (wes_different_names) {
  ids <- read_tsv(matching_IDs, show_col_types = FALSE)
  # replace the colnames corresponding to the secondary_ID with the panel_ID. Look up the corresponding name for each column in ids
  for (i in 2:ncol(wes)) {
    new_name <- ids$panel_ID[which(ids$secondary_ID == colnames(wes)[i])]
    new_name <- gsub("M0*", "M", new_name)
    new_name <- gsub("-", "_", new_name)
    colnames(wes)[i] <- new_name
  }
}

# remove columns whose names are NA
wes <- wes[, !is.na(colnames(wes))]

input_dir <- paste0(dir, "/../results/collection_of_specific_samples/")
list.files(input_dir, full.names = TRUE, pattern = "_gene_calls.tsv") -> files

# filter out genes from WES that are not in paneldesign

## in all files, save the common genes in a list
common_panel_genes <- c()
for (file in files) {
  read_tsv(file, show_col_types = FALSE) %>%
    arrange(gene) -> p
  if (file == files[1]) {
    common_panel_genes <- p$gene
  }
  common_panel_genes <- intersect(common_panel_genes, p$gene)
}

read_tsv(files[1], show_col_types = FALSE) %>%
  arrange(gene) -> panel
colnames(panel) <- rename_names(colnames(panel))
wes[, c(which(colnames(wes) %in% colnames(panel)))] %>%
  filter(gene %in% common_panel_genes) -> wes

# Does this set of 28 samples include samples to be discarded due to low high mapping quality and read depth?
samples_high_mapp_high_rd <- read_tsv(paste0(dir, "/../results/stats/samples_high_mapp_high_rd.tsv"))
colnames(panel) %in% samples_high_mapp_high_rd
# none -> good

panel %>%
  filter(!gene %in% wes$gene) %>%
  dplyr::select(gene) -> panel_gene_not_in_wes
# none -> good

contingency_list <- list()
TF_PN_list <- list()
sp_TF_PN_list <- list()

for (file in files) {
  read_tsv(file, show_col_types = FALSE) %>%
    arrange(gene) -> p
  colnames(p) <- rename_names(colnames(p))
  p %>% filter(gene %in% common_panel_genes) -> p

  if (!identical(p$gene, wes$gene)) {
    stop("Gene names are not identical")
  }

  p_filt <- p[, colnames(p) %in% colnames(wes)]
  wes_filt <- wes[, colnames(wes) %in% colnames(p)]
  p_filt %>% dplyr::select(order(colnames(p_filt), decreasing = TRUE)) -> p_filt
  wes_filt %>% dplyr::select(order(colnames(wes_filt), decreasing = TRUE)) -> wes_filt

  if (!identical(colnames(p_filt), colnames(wes_filt))) {
    stop("Sample names are not identical")
  }

  p_filt %>%
    select(-gene) %>%
    as.matrix() -> panel_cnv_calls

  wes_filt %>%
    select(-gene) %>%
    as.matrix() -> wes_cnv_calls

  rownames(panel_cnv_calls) <- p$gene
  rownames(wes_cnv_calls) <- wes$gene

  # in wes, convert all 2 to NO CHANGE, values > 2 to DELETION and values < 2 to AMPLIFICATION
  wes_cnv_calls[wes_cnv_calls > neutral] <- "AMPLIFICATION"
  wes_cnv_calls[wes_cnv_calls < neutral] <- "DELETION"
  wes_cnv_calls[wes_cnv_calls == neutral] <- "NO CHANGE"
  # in p, convert all NA to NO CHANGE
  panel_cnv_calls[is.na(panel_cnv_calls)] <- "NO CHANGE"
  contingency <- table(wes_cnv_calls, panel_cnv_calls)
  contingency_list[[gsub("_gene_calls.tsv", "", gsub(".*/purecn[_]*", "", file))]] <- contingency

  # Binary classification: CNV or not
  # It seems that the assumption that an amplification call in wes would not occur on the same gene as a deletion call in p, and vice versa, is not valid. Hence, we obmit this step

  # Compute TP, FP, TN, FN
  general <- wes_cnv_calls
  specific <- wes_cnv_calls
  for (i in 1:nrow(wes_cnv_calls)) {
    for (j in 1:ncol(wes_cnv_calls)) {
      # if both are NO CHANGE -> TN
      if (wes_cnv_calls[i, j] == "NO CHANGE" && panel_cnv_calls[i, j] == "NO CHANGE") {
        general[i, j] <- "TN"
        specific[i, j] <- "TN"
        # if wes is CNV but panel is not -> FN
      } else if (wes_cnv_calls[i, j] != "NO CHANGE" && panel_cnv_calls[i, j] == "NO CHANGE") {
        general[i, j] <- "FN"
        specific[i, j] <- paste("FN", wes_cnv_calls[i, j])
        # if wes is not CNV but panel is -> FP
      } else if (wes_cnv_calls[i, j] == "NO CHANGE" && panel_cnv_calls[i, j] != "NO CHANGE") {
        general[i, j] <- "FP"
        specific[i, j] <- paste("FP", panel_cnv_calls[i, j])
        # if both are CNV and same type of CNV -> TP
      } else if (wes_cnv_calls[i, j] != "NO CHANGE" && wes_cnv_calls[i, j] == panel_cnv_calls[i, j]) {
        general[i, j] <- "TP"
        specific[i, j] <- paste("TP", wes_cnv_calls[i, j])
        # if both are CNV but different types of CNV -> FP
      } else if (wes_cnv_calls[i, j] != "NO CHANGE" && wes_cnv_calls[i, j] != panel_cnv_calls[i, j]) {
        general[i, j] <- "FP"
        if (wes_cnv_calls[i, j] == "AMPLIFICATION") {
          specific[i, j] <- "FP AMPLIFICATION called but is DELETION"
        } else {
          specific[i, j] <- "FP DELETION called but is AMPLIFICATION"
        }
      }
    }
  }

  # do not use table(general) because in case of no TP, later row merging will screw up the results
  ct <-
    c(
      "TP" = sum(general == "TP"),
      "FP" = sum(general == "FP"),
      "TN" = sum(general == "TN"),
      "FN" = sum(general == "FN")
    )
  TF_PN_list[[gsub("_gene_calls.tsv", "", gsub(".*/purecn[_]*", "", file))]] <-
    ct

  # same with specific tables
  sp_ct <-
    c(
      "TP AMPLIFICATION" = sum(specific == "TP AMPLIFICATION"),
      "TP DELETION" = sum(specific == "TP DELETION"),
      "FP AMPLIFICATION" = sum(specific == "FP AMPLIFICATION"),
      "FP DELETION" = sum(specific == "FP DELETION"),
      "FN AMPLIFICATION" = sum(specific == "FN AMPLIFICATION"),
      "FN DELETION" = sum(specific == "FN DELETION"),
      "FP AMPLIFICATION called but is DELETION" = sum(specific == "FP AMPLIFICATION called but is DELETION"),
      "FP DELETION called but is AMPLIFICATION" = sum(specific == "FP DELETION called but is AMPLIFICATION"),
      "TN" = sum(specific == "TN")
    )
  sp_TF_PN_list[[gsub("_gene_calls.tsv", "", gsub(".*/purecn[_]*", "", file))]] <- sp_ct
}
```

## Number of genes with copy number (CN) changes in WES samples, among the 232 genes in the panel design, per sample

```{r}
apply(round(wes[-1], 0), 2, function(x) {
  length(which((x != 0)))
}) %>%
  sort(., decreasing = TRUE) %>%
  as.data.frame() %>%
  rownames_to_column() -> nb_cnvs_wes
colnames(nb_cnvs_wes) <- c("sample", "# gene CN changes")

apply(round(wes[-1], 0), 2, function(x) {
  length(which((x > 0)))
}) %>%
  sort(., decreasing = TRUE) %>%
  as.data.frame() %>%
  rownames_to_column() -> nb_amplifications_wes
colnames(nb_amplifications_wes) <-
  c("sample", "# gene amplifications")
apply(round(wes[-1], 0), 2, function(x) {
  length(which((x < 0)))
}) %>%
  sort(., decreasing = TRUE) %>%
  as.data.frame() %>%
  rownames_to_column() -> nb_deletions_wes
colnames(nb_deletions_wes) <- c("sample", "# gene deletions")

# plot both the number of deletions and amplifications in one plot
nb_amplifications_wes %>%
  left_join(nb_deletions_wes, by = "sample") %>%
  pivot_longer(
    cols = -sample,
    names_to = "type",
    values_to = "count"
  ) %>%
  ggplot(aes(
    x = reorder(sample, -count),
    y = count,
    fill = type
  )) +
  geom_bar(position = "stack", stat = "identity") +
  coord_flip() +
  labs(x = "Sample", y = "# gene CN changes in WES samples among the 232 genes covered by the panel")
```



## Example

```{r}
len <- length(contingency_list)
if (len > 5) {
  n <- len - 3 # choose almost the last one, but not hmm if possible
} else {
  n <- 1 # not many branches, just take the first one
}
names(contingency_list)[n]
contingency_list[[n]]
```


## Opposite CNV calls?

Is there an element in the contingency list that is non zero in the first row and second column or second row and first column?

If yes, then the assumption that *an amplification in WES would not be called on the same gene as a deletion in a panel, and vice versa*, is not valid.

If there is no output below, then are no such occurrences, thus the assumption is met. If there is an output below, the the assumption is not met. When computing false positives, TP and FN, one must check for the same type of CNV instead of only a binary classification of COPY NUMBER CHANGE or not.

```{r}
for (i in 1:length(contingency_list)) {
  try(
    expr = {
      if (contingency_list[[i]]["AMPLIFICATION", "DELETION"] != 0 |
        contingency_list[[i]]["DELETION", "AMPLIFICATION"] != 0) {
        print((contingency_list)[i])
      }
    },
    silent = TRUE
  )
}
```


# Summaries for all samples

~Note that many samples failed evaluation, especially when using no PON, as can be seen by a much lower sum (= number of genes x number of samples) in some rows.~

```{r amp and del separately}
# all sp_TF_PN rows into one table
sp_TF_PN <- do.call(rbind, sp_TF_PN_list)

# add precision and recall and sum columns
sp_TF_PN %>%
  as.data.frame() %>%
  rownames_to_column() %>%
  as_tibble() %>%
  rowwise() %>%
  mutate(
    sum = sum(
      `TP AMPLIFICATION`,
      `TP DELETION`,
      `FP AMPLIFICATION`,
      `FP DELETION`,
      `FN AMPLIFICATION`,
      `FN DELETION`,
      `FP AMPLIFICATION called but is DELETION`,
      `FP DELETION called but is AMPLIFICATION`,
      `TN`
    ),
    all = `TP AMPLIFICATION` + `TP DELETION` + `FP AMPLIFICATION` + `FP DELETION` + `FN AMPLIFICATION` + `FN DELETION` + `FP AMPLIFICATION called but is DELETION` + `FP DELETION called but is AMPLIFICATION` + `TN`,
    precision_amplification = round(
      `TP AMPLIFICATION` / (
        `TP AMPLIFICATION` + `FP AMPLIFICATION` + `FP AMPLIFICATION called but is DELETION`
      ),
      2
    ),
    recall_amplification = round(`TP AMPLIFICATION` / (`TP AMPLIFICATION` + `FN AMPLIFICATION`), 2),
    f1_amplification = round(
      2 * (precision_amplification * recall_amplification) / (precision_amplification + recall_amplification),
      2
    ),
    precision_deletion = round(
      `TP DELETION` / (
        `TP DELETION` + `FP DELETION` + `FP DELETION called but is AMPLIFICATION`
      ),
      2
    ),
    recall_deletion = round(`TP DELETION` / (`TP DELETION` + `FN DELETION`), 3),
    f1_deletion = round(
      2 * (precision_deletion * recall_deletion) / (precision_deletion + recall_deletion),
      2
    ),
    accuracy_amplification = round(
      (`TP AMPLIFICATION` + `TN`) / all,
      2
    ),
    accuracy_deletion = round(
      (`TP DELETION` + `TN`) / all,
      2
    ),
    # MCC = (TP * TN - FP * FN) / sqrt((TP + FP) * (TP + FN) * (TN + FP) * (TN + FN))
    # example (19 * 5745 - 452 * 14) / sqrt((19 + 452) * (19 + 14) * (5745 + 452) * (5745 + 14))
    MCC_amplification = round(
      (as.numeric(`TP AMPLIFICATION`) * as.numeric(`TN`) - (as.numeric(`FP AMPLIFICATION`) + as.numeric(`FP AMPLIFICATION called but is DELETION`)) * as.numeric(`FN AMPLIFICATION`)) /
        sqrt((as.numeric(`TP AMPLIFICATION`) + (as.numeric(`FP AMPLIFICATION`) + as.numeric(`FP AMPLIFICATION called but is DELETION`))) * (as.numeric(`TP AMPLIFICATION`) + as.numeric(`FN AMPLIFICATION`)) * (as.numeric(`TN`) + (as.numeric(`FP AMPLIFICATION`) + as.numeric(`FP AMPLIFICATION called but is DELETION`))) * (as.numeric(`TN`) + as.numeric(`FN AMPLIFICATION`))),
      2
    ),
    MCC_deletion = round(
      (as.numeric(`TP DELETION`) * as.numeric(`TN`) - (as.numeric(`FP DELETION`) + as.numeric(`FP DELETION called but is AMPLIFICATION`)) * as.numeric(`FN DELETION`)) /
        sqrt((as.numeric(`TP DELETION`) + (as.numeric(`FP DELETION`) + as.numeric(`FP DELETION called but is AMPLIFICATION`))) * (as.numeric(`TP DELETION`) + as.numeric(`FN DELETION`)) * (as.numeric(`TN`) + (as.numeric(`FP DELETION`) + as.numeric(`FP DELETION called but is AMPLIFICATION`))) * (as.numeric(`TN`) + as.numeric(`FN DELETION`))),
      2
    )
  ) -> sp_TF_PN_scores

colnames(sp_TF_PN_scores)[1] <- "branch"
```

## Amplifications

Recall = TPR = Sensitivity = TP / (TP + FN)  
Precision = PPV = TP / (TP + FP)  
F1 = 2 * (precision * recall) / (precision + recall)  
Accuracy = (TP + TN) / (TP + TN + FP + FN)  
AUC = 0.5 * (recall_amplification + recall_deletion)  
MCC = (TP * TN - FP * FN) / sqrt((TP + FP) * (TP + FN) * (TN + FP) * (TN + FN))  

```{r compute amp}
scores_amp <- sp_TF_PN_scores[, c(
  "branch",
  "TP AMPLIFICATION",
  "TN",
  "FP AMPLIFICATION",
  "FN AMPLIFICATION",
  "precision_amplification",
  "recall_amplification",
  "f1_amplification",
  "accuracy_amplification",
  "MCC_amplification"
)]
colnames(scores_amp) <- gsub("AMPLIFICATION", "", colnames(scores_amp))
colnames(scores_amp) <- gsub("_amplification", "", colnames(scores_amp))
scores_amp %>% kable()
```

```{r, layout="l-page",fig.width=15, fig.height=5}
scores_amp %>%
  mutate(
    branch = ifelse(!grepl("pon", branch), paste0("no_pon_", branch), branch),
    method = gsub("^.*pon_", "", branch),
    pon = ifelse(grepl("no_pon", branch), "none", gsub("_pon_.*", "", branch))
  ) %>%
  pivot_longer(
    cols = c(precision, recall, f1),
    names_to = "metric",
    values_to = "value"
  ) %>%
  ggplot(aes(x = pon, y = value, fill = method)) +
  geom_bar(position = "dodge", stat = "identity") +
  geom_text(aes(label = round(value, 2)), position = position_dodge(width = 0.9), vjust = -0.25, size = 2) +
  facet_wrap(~metric, scales = "fixed") +
  labs(x = "PON", y = "value")
```

## Deletions

```{r compute del}
scores_del <- sp_TF_PN_scores[, c(
  "branch",
  "TP DELETION",
  "TN",
  "FP DELETION",
  "FN DELETION",
  "precision_deletion",
  "recall_deletion",
  "f1_deletion",
  "accuracy_deletion",
  "MCC_deletion"
)]
colnames(scores_del) <- gsub("DELETION", "", colnames(scores_del))
colnames(scores_del) <- gsub("_deletion", "", colnames(scores_del))
scores_del %>% kable()
```

```{r, layout="l-page",fig.width=15, fig.height=5}
scores_del %>%
  mutate(
    branch = ifelse(!grepl("pon", branch), paste0("no_pon_", branch), branch),
    method = gsub("^.*pon_", "", branch),
    pon = ifelse(grepl("no_pon", branch), "none", gsub("_pon_.*", "", branch))
  ) %>%
  pivot_longer(
    cols = c(precision, recall, f1),
    names_to = "metric",
    values_to = "value"
  ) %>%
  ggplot(aes(x = pon, y = value, fill = method)) +
  geom_bar(position = "dodge", stat = "identity") +
  geom_text(aes(label = round(value, 2)), position = position_dodge(width = 0.9), vjust = -0.25, size = 2) +
  facet_wrap(~metric, scales = "fixed") +
  labs(x = "PON", y = "value")
```



# Scatterplots to compare final LFC values and diagnosize source(s) of differences between results

```{r extract gene coordinates, include = FALSE, eval = TRUE}
list.files(paste0(dir, "/../results/collection_of_specific_samples/"), pattern = "_genes.csv", full.names = TRUE, recursive = TRUE)[1] %>%
  read_csv() %>%
  filter(gene.symbol %in% wes_all$gene) %>%
  select(gene.symbol, chr, start, end) %>%
  {
    write_tsv(., paste0(dir, "/../resources/paneldesign/gene_coordinates.tsv"))
  } -> gene_coordinates
```


LFC = log2 fold change
adj. = adjusted for tumor ploidy and purity

```{r}
#--- Create different scatterplots for one sample ---#

scatterplots_for_sample <- function(
    panel_sample_id,
    wes_sample_id = NULL, # look up if not provided,
    branch,
    id_lookup_table = matching_IDs,
    gene_coordinates_tsv = paste0(dir, "/../resources/paneldesign/gene_coordinates.tsv"),
    panels_genes_csv = NULL,
    wes_dnacopy_seg = paste0(dir, "/../results/collection_of_specific_samples/wes/bwa.copywriter.", wes_sample_id, "-T1-DNA1-WES1_segments.txt"),
    cnr_file = paste0(dir, "/../results/collection_of_specific_samples/wes/bwa_mem2.cnvkit.", wes_sample_id, "-T1-DNA1-WES1.cnr"),
    panel_rds = NULL,
    panel_cnvs_file = paste0(dir, "/../results/collection_of_specific_samples/", gsub("/", "_", branch), "_gene_calls.tsv"),
    wes_cnvs_file = wes_gene_calls,
    include_cnr = FALSE,
    include_non_adj = FALSE) {
  # Look up wes_sample_id if not provided
  if (is.null(wes_sample_id)) {
    read_tsv(id_lookup_table, show_col_types = FALSE) %>%
      filter(rename_names(panel_ID) == panel_sample_id) %>%
      dplyr::select(secondary_ID) %>%
      as.character() -> wes_sample_id

    # find files in the folder paste0(dir, "/../resources/gold_standard/") that contain wes_sample_id
    matching_files <- list.files(paste0(dir, "/../resources/gold_standard/"), pattern = wes_sample_id, full.names = TRUE)
    if (length(matching_files) == 0) {
      stop("No matching files found for wes_sample_id")
    }

    wes_dnacopy_seg <- grep("dnacopy", matching_files, value = TRUE)
    if (length(wes_dnacopy_seg) < 1) {
      wes_dnacopy_seg <- grep("segments", matching_files, value = TRUE)
    }
    cnr_file <- grep("cnr$", matching_files, value = TRUE)
  }

  files <- list.files(paste0(dir, "/../results/collection_of_specific_samples/"), full.names = TRUE, recursive = TRUE)
  files %>%
    # add a column with row indices
    tibble(file = ., row = 1:length(.)) %>%
    mutate(file = rename_names(file)) %>%
    filter(
      grepl(panel_sample_id, file),
      grepl(branch, file)
    ) -> matching_p_files

  matching_p_files %>%
    filter(grepl("_genes.csv", file)) %>%
    pull(row) -> genes_csv_index

  panels_genes_csv <- files[genes_csv_index]

  matching_p_files %>%
    filter(grepl(".rds", file)) %>%
    pull(row) -> rds_index

  panel_rds <- files[rds_index]

  # find the intersection of the names of panel_sample_id and panel_rds
  panel_rds %>%
    basename() %>%
    str_remove(".rds") -> other_panel_sample_id

  # Read gene coordinates
  gene_coordinates <- read_tsv(gene_coordinates_tsv, show_col_types = FALSE)
  gene_coordinates_gr <- GRanges(
    seqnames = gene_coordinates$chr,
    ranges = IRanges(start = gene_coordinates$start, end = gene_coordinates$end, names = gene_coordinates$gene.symbol, genes = gene_coordinates$gene.symbol)
  )

  # Read panel C data for the 232 genes
  panels_segmean_C <- read_csv(panels_genes_csv, show_col_types = FALSE) %>%
    filter(!grepl("chr[0-9x]*:", gene.symbol), !grepl("Antitarget", gene.symbol)) %>%
    mutate(p_segmean = seg.mean) %>%
    dplyr::select(gene.symbol, C, p_segmean, type)

  # Read wes_seg_mean data
  w <- read_tsv(wes_dnacopy_seg, show_col_types = FALSE)
  wes_gr <- GRanges(
    seqnames = w$chrom,
    ranges = IRanges(start = w$loc.start, end = w$loc.end),
    wes_seg_mean = w$seg.mean
  )

  # Intersect wes wes_seg_mean with gene coordinates
  overlaps <- findOverlaps(wes_gr, gene_coordinates_gr)
  contained_genes <- gene_coordinates_gr[subjectHits(overlaps)]
  mcols(contained_genes)$wes_seg_mean <- mcols(wes_gr[queryHits(overlaps)])$wes_seg_mean
  wes_gene_segmean <- data.frame(
    "genes" = contained_genes$genes,
    "wes_seg_mean" = contained_genes$wes_seg_mean
  )

  # Merge with panels_segmean_C
  df_p_segmean_C_wes_segmean <- panels_segmean_C %>%
    full_join(wes_gene_segmean, by = c("gene.symbol" = "genes"))

  # Read adjusted wes_seg_mean data from panel
  rds <- readRDS(panel_rds)
  r <- rds$results[[1]]
  R <- r$seg$seg.mean
  a <- r$purity
  T <- r$ploidy
  r$seg$p_seg_mean_adjusted <- (a * T * R + 2 * (1 - a) * R - 2 * (1 - a)) / (a * T)
  rseg_gr <- GRanges(
    seqnames = r$seg$chrom,
    ranges = IRanges(start = r$seg$loc.start, end = r$seg$loc.end),
    strand = "*",
    p_seg_mean_adjusted = r$seg$p_seg_mean_adjusted
  )

  # Intersect adjusted wes_seg_mean with gene coordinates
  overlaps <- findOverlaps(rseg_gr, gene_coordinates_gr)
  contained_genes <- gene_coordinates_gr[subjectHits(overlaps)]
  mcols(contained_genes)$p_seg_mean_adjusted <- mcols(rseg_gr[queryHits(overlaps)])$p_seg_mean_adjusted
  p_rsl <- data.frame(
    "genes" = contained_genes$genes,
    "p_seg_mean_adjusted" = contained_genes$p_seg_mean_adjusted
  )

  # Merge with panels_segmean_C
  df_p_segmean_adj_C_wes <- wes_gene_segmean %>%
    full_join(p_rsl,
      by = c("genes" = "genes"), relationship =
        "many-to-many"
    ) %>%
    full_join(panels_segmean_C, by = c("genes" = "gene.symbol"))

  # Read panel and wes cnv data
  read_tsv(panel_cnvs_file, show_col_types = FALSE) %>%
    arrange(gene) -> panel_cnvs
  # read_tsv(wes_gene_calls, col_names = TRUE, show_col_types = FALSE) %>%
  #  arrange(genes) -> wes_cnvs
  wes_cnvs <- wes

  # Extract column for the given sample
  panel_cnvs_sample <- panel_cnvs %>%
    dplyr::select(gene, other_panel_sample_id)
  colnames(panel_cnvs_sample)[2] <- "panel_cnvs"
  wes_cnvs_sample <- wes_cnvs %>%
    rename(
      gene = gene,
      sample := 2
    ) %>%
    #    dplyr::select(genes, wes_sample_id)
    dplyr::select(gene, sample)
  colnames(wes_cnvs_sample)[2] <- "wes_cnvs"

  # Merge with panels_segmean_C
  p_wes_cnvs_sample <- df_p_segmean_adj_C_wes %>%
    full_join(panel_cnvs_sample, by = c("genes" = "gene")) %>%
    full_join(wes_cnvs_sample, by = c("genes" = "gene"))

  # Add a column type based on cnv values
  p_wes_cnvs_sample$type <- NULL

  # replace all NAs with 0
  p_wes_cnvs_sample$wes_cnvs[is.na(p_wes_cnvs_sample$wes_cnvs)] <- 0

  for (row in 1:nrow(p_wes_cnvs_sample)) {
    if (is.na(p_wes_cnvs_sample$panel_cnvs[row])) {
      if (p_wes_cnvs_sample$wes_cnvs[row] == neutral) {
        p_wes_cnvs_sample$type[row] <- "TN"
      } else if (p_wes_cnvs_sample$wes_cnvs[row] > neutral) {
        p_wes_cnvs_sample$type[row] <- "FN amplification"
      } else if (p_wes_cnvs_sample$wes_cnvs[row] < neutral) {
        p_wes_cnvs_sample$type[row] <- "FN deletion"
      }
    } else if (p_wes_cnvs_sample$panel_cnvs[row] == "AMPLIFICATION") {
      if (p_wes_cnvs_sample$wes_cnvs[row] == neutral) {
        p_wes_cnvs_sample$type[row] <- "FP amplification"
      } else if (p_wes_cnvs_sample$wes_cnvs[row] > neutral) {
        p_wes_cnvs_sample$type[row] <- "TP amplification"
      } else if (p_wes_cnvs_sample$wes_cnvs[row] < neutral) {
        p_wes_cnvs_sample$type[row] <- "FP detected amplification but is deletion"
      }
    } else if (p_wes_cnvs_sample$panel_cnvs[row] == "DELETION") {
      if (p_wes_cnvs_sample$wes_cnvs[row] == neutral) {
        p_wes_cnvs_sample$type[row] <- "FP deletion"
      } else if (p_wes_cnvs_sample$wes_cnvs[row] > neutral) {
        p_wes_cnvs_sample$type[row] <- "FP detected deletion but is amplification"
      } else if (p_wes_cnvs_sample$wes_cnvs[row] < neutral) {
        p_wes_cnvs_sample$type[row] <- "TP deletion"
      }
    }
  }

  # check cnv calls
  # print(unique(p_wes_cnvs_sample$wes_cnvs))
  # print(unique(p_wes_cnvs_sample$panel_cnvs))
  # print(unique(p_wes_cnvs_sample$type))

  # Plots
  plots <- list()

  # Scatterplot wes_seg_mean from dnacopy.seg from wes on the x-axis, adjusted wes_seg_mean from panel on the y-axis
  plots[[1]] <- ggplot(p_wes_cnvs_sample, aes(x = wes_seg_mean, y = p_seg_mean_adjusted, color = type)) +
    geom_smooth(aes(group = 1), method = "lm", colour = "grey") +
    geom_point() +
    theme_minimal() +
    labs(x = "WES: mean segment LFC (not adj.)", y = "Panel: mean segment LFC (adj.)") +
    scale_color_manual(values = c(
      "TN" = "grey",
      "TP amplification" = "green",
      "TP deletion" = "darkgreen",
      "FN amplification" = "tomato",
      "FN deletion" = "cyan",
      "FP detected amplification but is deletion" = "red",
      "FP deletion" = "orange",
      "FP amplification" = "orange3",
      "FP detected deletion but is amplification" = "darkred"
    )) +
    theme(legend.position = "bottom")


  # Scatterplot wes_seg_mean from dnacopy.seg from wes on the x-axis, C from panel on the y-axis
  plots[[2]] <- ggplot(p_wes_cnvs_sample, aes(x = wes_seg_mean, y = C, color = type)) +
    geom_smooth(aes(group = 1), method = "lm", colour = "grey") +
    geom_point() +
    theme_minimal() +
    labs(x = "WES: mean segment LFC (not adj.)", y = "Panel: absolute copy numbers (C)") +
    scale_color_manual(values = c(
      "TN" = "grey",
      "TP amplification" = "green",
      "TP deletion" = "darkgreen",
      "FN amplification" = "tomato",
      "FN deletion" = "cyan",
      "FP detected amplification but is deletion" = "red",
      "FP deletion" = "orange",
      "FP amplification" = "orange3",
      "FP detected deletion but is amplification" = "darkred"
    )) +
    scale_y_continuous(breaks = function(x) unique(floor(pretty(seq(0, (max(x) + 1) * 1.1))))) +
    theme(legend.position = "none")

  if (include_non_adj) {
    # Scatterplot wes_seg_mean from dnacopy.seg from wes on the x-axis, wes_seg_mean from panel on the y-axis
    plots[[3]] <- ggplot(p_wes_cnvs_sample, aes(x = wes_seg_mean, y = p_segmean, color = type)) +
      geom_smooth(aes(group = 1), method = "lm", colour = "grey") +
      geom_point() +
      theme_minimal() +
      labs(x = "WES: mean segment LFC (not adj.)", y = "Panel: mean segment LFC (not adj.)") +
      scale_color_manual(values = c(
        "TN" = "grey",
        "TP amplification" = "green",
        "TP deletion" = "darkgreen",
        "FN amplification" = "tomato",
        "FN deletion" = "cyan",
        "FP detected amplification but is deletion" = "red",
        "FP deletion" = "orange",
        "FP amplification" = "orange3",
        "FP detected deletion but is amplification" = "darkred"
      )) +
      theme(legend.position = "bottom")
  }

  if (include_cnr) {
    # Read cnr data from wes
    cnr <- read_tsv(cnr_file, show_col_types = FALSE)
    cnr_gr <- GRanges(
      seqnames = cnr$chromosome,
      ranges = IRanges(start = cnr$start, end = cnr$end),
      strand = "*",
      wes_seg_mean_cnr = cnr$log2
    )

    # Intersect cnr with gene coordinates
    overlaps <- findOverlaps(cnr_gr, gene_coordinates_gr)
    contained_genes <- gene_coordinates_gr[subjectHits(overlaps)]
    mcols(contained_genes)$wes_seg_mean_cnr <- mcols(cnr_gr[queryHits(overlaps)])$wes_seg_mean_cnr
    wes_gene_segmean_cnr <- data.frame(
      "genes" = contained_genes$genes,
      "wes_seg_mean_cnr" = contained_genes$wes_seg_mean_cnr
    )

    # Merge with panels_segmean_C
    df_p_segmean_C_wes_cnr <- df_p_segmean_C_wes_segmean %>%
      full_join(wes_gene_segmean_cnr, by = c("gene.symbol" = "genes"), relationship = "many-to-many")

    # full_join with wes_gene_segmean_cnr with cnvs
    df_wes_cnr_cnvs_sample <- p_wes_cnvs_sample %>%
      full_join(wes_gene_segmean_cnr, by = c("genes" = "genes"), relationship = "many-to-many")

    # Scatterplot wes_seg_mean from cnr from wes on the x-axis, C from panel on the y-axis
    plots[[4]] <- ggplot(df_wes_cnr_cnvs_sample, aes(x = wes_seg_mean_cnr, y = p_seg_mean_adjusted, color = type)) +
      geom_smooth(aes(group = 1), method = "lm", colour = "grey") +
      geom_point() +
      theme_minimal() +
      labs(x = "WES bin log2 values (ca. 100 bins per gen)", y = "Panel: mean segment LFC (adj.)") +
      scale_color_manual(values = c(
        "TN" = "grey",
        "TP amplification" = "green",
        "TP deletion" = "darkgreen",
        "FN amplification" = "tomato",
        "FN deletion" = "cyan",
        "FP detected amplification but is deletion" = "red",
        "FP deletion" = "orange",
        "FP amplification" = "orange3",
        "FP detected deletion but is amplification" = "darkred"
      )) +
      theme(legend.position = "bottom")
  }

  return(plots)
}
```


## All 4 plots for one example sample and branch (cbs & Hclust)


First, find a sample with multiple CNV calls of every type in WES results. Let us filter for those with a lot of copy number changes.

```{r}
# calls = read_tsv(wes_gene_calls)
calls <- wes
total_calls <- cbind(
  apply(calls[, -1], 2, function(x) sum(x > neutral)),
  apply(calls[, -1], 2, function(x) sum(x < neutral))
)

colnames(total_calls) <- c("amplifications", "deletions")
total_calls %>%
  as.data.frame() %>%
  rownames_to_column() %>%
  as_tibble() -> samples_with_deletions

samples_with_deletions %>% kable()

# choose one of the samples
samples_with_deletions %>%
  mutate(score = amplifications * deletions) %>%
  arrange(desc(score)) %>%
  dplyr::select(rowname) -> scores

# with most calls
scores[1, 1] %>% pull() -> highest
# with medium nb of calls
scores[nrow(scores) %/% 2, 1] %>% pull() -> medium
# with least calls
scores[nrow(scores), 1] %>% pull() -> lowest

cat("Let us first choose sample", highest)
```


### PON: None, Branch: CBS x Hclust

#### Sample `r highest`

A sample with many CN calls.

```{r, layout="l-page",fig.width=12, fig.height=5, warning=FALSE, message=FALSE}
p <- scatterplots_for_sample(
  panel_sample_id = highest,
  id_lookup_table = matching_IDs,
  branch = "purecn/cbs_Hclust"
)
grid.arrange(p[[1]], p[[2]], ncol = 2)
```

#### Sample `r medium`

A sample with around median number of CN calls.

```{r, layout="l-page",fig.width=12, fig.height=5, warning=FALSE, message=FALSE}
p <- scatterplots_for_sample(
  panel_sample_id = medium,
  id_lookup_table = matching_IDs,
  branch = "purecn/cbs_Hclust"
)
grid.arrange(p[[1]], p[[2]], ncol = 2)
```

#### Sample `r lowest`

A sample with few CN calls.

```{r, layout="l-page",fig.width=12, fig.height=5, warning=FALSE, message=FALSE}
p <- scatterplots_for_sample(
  panel_sample_id = lowest,
  id_lookup_table = matching_IDs,
  branch = "purecn/cbs_Hclust"
)
grid.arrange(p[[1]], p[[2]], ncol = 2)
```
